### PR TITLE
feat: Add multiple phonePrefix

### PR DESCRIPTION
**Pull request by jakiuncle**
Fix: https://github.com/casdoor/casdoor/issues/1397

### CODE REVIEW

File: object/permission.go [hunk 1]
```diff
@@ -296,6 +296,29 @@ func MigratePermissionRule() {
 	}
 }
 
+func MigratePhonePrefix() {
```
Comment (leo220yuyaodog) [inline object/permission.go:299]: Create a new file named migrate is better.

File: controllers/verification.go [hunk 2]
```diff
@@ -142,7 +143,11 @@ func (c *ApiController) SendVerificationCode() {
 			return
 		}
 
-		dest = fmt.Sprintf("+%s%s", organization.PhonePrefix, dest)
+		if (phonePrefix == "" || phonePrefix == "undefined") && userByPhone != nil {
```
Comment (leo220yuyaodog) [inline controllers/verification.go:146]: Handle undefine in frontend. Just has  "" condition in backend.

File: object/user.go [hunk 3]
```diff
@@ -47,6 +47,7 @@ type User struct {
 	Email             string   `xorm:"varchar(100) index" json:"email"`
 	EmailVerified     bool     `json:"emailVerified"`
 	Phone             string   `xorm:"varchar(100) index" json:"phone"`
+	PhonePrefix       string   `xorm:"varchar(100)" json:"phonePrefix"`
```
Comment (leo220yuyaodog) [inline object/user.go:50]: https://stackoverflow.com/questions/8284647/how-to-store-phone-numbers-on-mysql-databases
Both phone prefix and number save as a varchar maybe better. Formatting should be done in the view.

File: object/user.go [hunk 4]
```diff
@@ -177,6 +178,7 @@ type Userinfo struct {
 	Avatar      string `json:"picture,omitempty"`
 	Address     string `json:"address,omitempty"`
 	Phone       string `json:"phone,omitempty"`
+	PhonePrefix string `json:"phonePrefix,omitempty"`
```
Comment (leo220yuyaodog) [inline object/user.go:181]: Not need. The `Userinfo` struct  is for oidc protocol.

File: object/user.go (hunk 3 above)
Comment (hsluoyz) [reply object/user.go:50]: @leo220yuyaodog one answer sugggested storing country code separately

image

File: controllers/verification.go [hunk 5]
```diff
@@ -177,7 +184,7 @@ func (c *ApiController) ResetEmailOrPhone() {
 	checkDest := dest
 	organization := object.GetOrganizationByUser(user)
 	if destType == "phone" {
-		if object.HasUserByField(user.Owner, "phone", user.Phone) {
+		if object.HasUserByField(user.Owner, "phone", dest) {
```
Comment (hsluoyz) [inline controllers/verification.go:187]: Why change it?

File: controllers/auth.go [hunk 6]
```diff
@@ -240,6 +240,7 @@ func (c *ApiController) Login() {
 		var user *object.User
 		var msg string
 
+		// Login with code
```
Comment (hsluoyz) [inline controllers/auth.go:243]: Remove this

File: controllers/account.go [hunk 7]
```diff
@@ -137,7 +137,7 @@ func (c *ApiController) Signup() {
 
 	var checkPhone string
 	if application.IsSignupItemVisible("Phone") && form.Phone != "" {
-		checkPhone = fmt.Sprintf("+%s%s", form.PhonePrefix, form.Phone)
+		checkPhone = fmt.Sprintf("%s%s", form.PhonePrefix, form.Phone)
```
Comment (hsluoyz) [inline controllers/account.go:140]: Why remove + ?

File: controllers/auth.go [hunk 8]
```diff
@@ -257,16 +258,13 @@ func (c *ApiController) Login() {
 				}
 				checkDest = form.Username
 			} else {
-				verificationCodeType = "phone"
```
Comment (hsluoyz) [inline controllers/auth.go:260]: Why remove this?

File: web/src/ResetModal.js [hunk 9]
```diff
@@ -97,7 +101,7 @@ export const ResetModal = (props) => {
               textBefore={i18next.t("code:Code You Received")}
               onChange={setCode}
               method={"reset"}
-              onButtonClickArgs={[dest, destType, Setting.getApplicationName(application)]}
+              onButtonClickArgs={[dest, destType, Setting.getApplicationName(application), null, phonePrefix]}
```
Comment (hsluoyz) [inline web/src/ResetModal.js:104]: What's null?

File: web/src/SelectPhonePrefixBox.js [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+  }
+
+  render() {
+    const isOrgnazition = this.props.isOrgnazition;
+    const wid = isOrgnazition ? "100%" : 100;
+    let phoneArray = [];
+    if (isOrgnazition) {
+      phoneArray = Setting.CountryPrefiexphone;
+    } else {
+      for (let i = 0; i < this.props.obj.phonePrefix.length; i++) {
```
Comment (hsluoyz) [inline web/src/SelectPhonePrefixBox.js:42]: foreach

File: web/src/SelectPhonePrefixBox.js [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+  }
+
+  onChange(e) {
+    this.props.onChange(e);
+    this.setState({value: e});
+  }
+
+  render() {
+    const isOrgnazition = this.props.isOrgnazition;
+    const wid = isOrgnazition ? "100%" : 100;
```
Comment (hsluoyz) [inline web/src/SelectPhonePrefixBox.js:37]: wid?

File: web/src/Setting.js [hunk 12]
```diff
@@ -204,7 +207,7 @@ export const OtherProviderInfo = {
   },
 };
 
-export function getCountryRegionData() {
+export function getCountryRegionData(key) {
```
Comment (hsluoyz) [inline web/src/Setting.js:210]: Don't mess with getCountryRegionData

File: web/src/auth/LoginPage.js [hunk 13]
```diff
@@ -189,7 +189,7 @@ class LoginPage extends React.Component {
     } else {
       values["type"] = this.state.type;
     }
-    values["phonePrefix"] = this.getApplicationObj()?.organizationObj.phonePrefix;
+    // values["phonePrefix"] = this.getApplicationObj()?.organizationObj.phonePrefix[0];
```
Comment (hsluoyz) [inline web/src/auth/LoginPage.js:192]: Just remove

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Comment by hsluoyz** [comment]
Replaced by: https://github.com/casdoor/casdoor/pull/1557

**Pull request closed without merging**